### PR TITLE
Compaction stats

### PR DESCRIPTION
**Pull request by sitaram-kalluri**
**- What I did** Captures the commit log and access log compaction metrics.

**- How I did it**

**- How to verify it**

**- Description for the changelog**

### CODE REVIEW

File: at_secondary/at_secondary_server/lib/src/server/at_compaction_observer_impl.dart [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    // If key is found, retrieves the previousRun value and add to compactionStats;
+    // Else inserts a new key.
+    keyStore!.get(_getKey()).catchError((onError) {
+      _logger.finer('${_getKey()} not found in keystore. Inserting a new key.');
+    }).then((atData) => _insertInitialStats(atData));
+  }
+
+  void _insertInitialStats(atData) {
+    var compactionStats = CompactionStats()
+      ..nextRun = DateTime.now()
```
Comment (VJag) [inline at_secondary/at_secondary_server/lib/src/server/at_compaction_observer_impl.dart:33]: Does the compaction job honor this nextRun time? i.e Is the computed next run date time and actual run date equal?

File: at_secondary/at_secondary_server/lib/src/server/at_compaction_observer_impl.dart [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+      ..previousRun =
+          DateTime.fromMillisecondsSinceEpoch(compactionEndTimeInEpoch)
+      ..duration = DateTime.fromMillisecondsSinceEpoch(compactionEndTimeInEpoch)
+          .difference(
+              DateTime.fromMillisecondsSinceEpoch(compactionStartTimeInEpoch))
+      ..keysBeforeCompaction = sizeBeforeCompaction
+      ..keysAfterCompaction = atLogType.getSize()
+      ..nextRun = DateTime.fromMillisecondsSinceEpoch(compactionEndTimeInEpoch)
+          .add(Duration(minutes: _getCompactionFrequencyMins()));
+    await keyStore!
```
Comment (VJag) [inline at_secondary/at_secondary_server/lib/src/server/at_compaction_observer_impl.dart]: How should we handle any exception in start() or end() methods?

File: at_secondary/at_secondary_server/lib/src/server/at_compaction_observer_impl.dart [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        .put(_getKey(), AtData()..data = jsonEncode(compactionStats));
+  }
+
+  String _getKey() {
+    if (atLogType is AtCommitLog) {
+      return commitLogCompactionKey;
+    }
+    if (atLogType is AtAccessLog) {
+      return accessLogCompactionKey;
```
Comment (VJag) [inline at_secondary/at_secondary_server/lib/src/server/at_compaction_observer_impl.dart:92]: I reckon these are private: or _ keys

File: at_secondary/at_secondary_server/lib/src/verb/metrics/metrics_impl.dart [hunk 4]
```diff
@@ -275,3 +276,59 @@ class LastPkamMetricImpl implements MetricProvider {
...
+  }
+
+  @override
+  getMetrics({String? regex}) async {
+    var keyStore = SecondaryPersistenceStoreFactory.getInstance()
+        .getSecondaryPersistenceStore(
+            AtSecondaryServerImpl.getInstance().currentAtSign)!
+        .getSecondaryKeyStore();
+    var atData = await keyStore!.get(commitLogCompactionKey);
+    if (atData != null && atData.data != null) {
```
Comment (VJag) [inline at_secondary/at_secondary_server/lib/src/verb/metrics/metrics_impl.dart]: For whatever reason, if the stats are null (atData = null) what should we return. I suggest we return a valid empty JSON.

File: at_secondary/at_secondary_server/lib/src/verb/metrics/metrics_impl.dart [hunk 5]
```diff
@@ -275,3 +276,59 @@ class LastPkamMetricImpl implements MetricProvider {
...
+  }
+
+  @override
+  getMetrics({String? regex}) async {
+    var keyStore = SecondaryPersistenceStoreFactory.getInstance()
+        .getSecondaryPersistenceStore(
+            AtSecondaryServerImpl.getInstance().currentAtSign)!
+        .getSecondaryKeyStore();
+    var atData = await keyStore!.get(accessLogCompactionKey);
+    if (atData != null && atData.data != null) {
```
Comment (VJag) [inline at_secondary/at_secondary_server/lib/src/verb/metrics/metrics_impl.dart]: For whatever reason, if the stats are null (atData = null) what should we return. I suggest we return a valid empty JSON.

File: at_secondary/at_secondary_server/lib/src/server/at_compaction_observer_impl.dart (hunk 3 above)
Comment (sitaram-kalluri) [reply at_secondary/at_secondary_server/lib/src/server/at_compaction_observer_impl.dart:92]: yes Jagan, these are the private keys. Below are the keys.
const String commitLogCompactionKey = 'privatekey:commitLogCompactionStats';
const String accessLogCompactionKey = 'privatekey:accessLogCompactionStats';

File: at_secondary/at_secondary_server/lib/src/server/at_compaction_observer_impl.dart (hunk 1 above)
Comment (sitaram-kalluri) [reply at_secondary/at_secondary_server/lib/src/server/at_compaction_observer_impl.dart:33]: For the nextrun, in the end method, we add the compaction frequency mins to previous run DateTime.

File: at_secondary/at_secondary_server/lib/src/server/at_compaction_observer_impl.dart (hunk 2 above)
Comment (sitaram-kalluri) [reply at_secondary/at_secondary_server/lib/src/server/at_compaction_observer_impl.dart]: Added generic try-catch to handle exception

**Review by VJag — APPROVED** [review]

**Comment by sitaram-kalluri** [comment]
Changes are no longer valid. Hence closing the PR.

**Pull request closed without merging**